File: 07_linked_list/07_06_palindrome_linked_list/palindrome_linked_list.js
class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

function isPalindrome(head) {
    if (!head || !head.next) {
        return true;
    }

    let slow = head;
    let fast = head;
    while (fast.next && fast.next.next) {
        slow = slow.next;
        fast = fast.next.next;
    }

    let reversedHead = reverse(slow.next);
    let first = head;
    let second = reversedHead;
    while (second) {
        if (first.val !== second.val) {
            reverse(reversedHead);
            return false;
        }
        first = first.next;
        second = second.next;
    }

    reverse(reversedHead);
    return true;
}

function reverse(head) {
    let prev = null;
    let current = head;
    let nextNode = head;
    while (nextNode) {
        nextNode = nextNode.next;
        current.next = prev;
        prev = current;
        current = nextNode;
    }
    return prev;
}

// Test Case 1: Palindrome with even number of nodes
const head1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(3, new ListNode(2, new ListNode(1))))));
console.log("Test Case 1:", isPalindrome(head1));  // Output: true

// Test Case 2: Palindrome with odd number of nodes
const head2 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(2, new ListNode(1)))));
console.log("Test Case 2:", isPalindrome(head2));  // Output: true

// Test Case 3: Not a palindrome
const head3 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
console.log("Test Case 3:", isPalindrome(head3));  // Output: false

// Test Case 4: Single node
const head4 = new ListNode(1);
console.log("Test Case 4:", isPalindrome(head4));  // Output: true

// Test Case 5: Empty list
const head5 = null;
console.log("Test Case 5:", isPalindrome(head5));  // Output: true

File: 07_linked_list/07_06_palindrome_linked_list/palindrome_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def is_palindrome(head):
    if not head or not head.next:
        return True

    slow = head
    fast = head
    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next

    reversed_head = reverse(slow.next)
    first = head
    second = reversed_head
    while second:
        if first.val != second.val:
            reverse(reversed_head)
            return False
        first = first.next
        second = second.next

    reverse(reversed_head)
    return True

def reverse(head):
    prev = None
    current = head
    next_node = head
    while next_node:
        next_node = next_node.next
        current.next = prev
        prev = current
        current = next_node
    return prev

# Test Case 1: Palindrome with even number of nodes
head1 = ListNode(1, ListNode(2, ListNode(3, ListNode(3, ListNode(2, ListNode(1))))))
print("Test Case 1:", is_palindrome(head1))  # Output: True

# Test Case 2: Palindrome with odd number of nodes
head2 = ListNode(1, ListNode(2, ListNode(3, ListNode(2, ListNode(1)))))
print("Test Case 2:", is_palindrome(head2))  # Output: True

# Test Case 3: Not a palindrome
head3 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
print("Test Case 3:", is_palindrome(head3))  # Output: False

# Test Case 4: Single node
head4 = ListNode(1)
print("Test Case 4:", is_palindrome(head4))  # Output: True

# Test Case 5: Empty list
head5 = None
print("Test Case 5:", is_palindrome(head5))  # Output: True
